corehq/apps/hqwebapp/models.py

```diff
@@ -53,3 +52,45 @@ def get_latest_alert(cls):
             return active_alerts[0]
         else:
             return ''
+
+
+class UserAgent(models.Model):
+    MAX_LENGTH = 255
+
+    value = models.CharField(max_length=MAX_LENGTH, db_index=True)
+
+
+class UserAccessLogManager(models.Manager):
+    def create(self, **obj_data):
+        user_agent = obj_data.pop('user_agent', '')
+        if user_agent:
+            user_agent = user_agent[:UserAgent.MAX_LENGTH]
+            agent_ref, _ = UserAgent.objects.get_or_create(value=user_agent)
```

Comment:
[`get_or_create`](https://github.com/django/django/blob/d4ac23bee1c84d8e4610350202ac068fc90f38c0/django/db/models/query.py#L571) does a `SELECT` and then an `INSERT` if that fails. The [solution](https://stackoverflow.com/a/13210391/10840) I had linked earlier, which suggested hashing the value and using the hash as the foreign key, could be implemented with a single `INSERT OR IGNORE`. That seems like a good idea for something as high-volume as this. Another nice aspect of that solution is it does not require an index on the full user agent value (only the hash/key must be indexed).

Thoughts?

Reply:
I don't think it's worth the complexity. I agree with your initial point about moving them off to a separate table to handle the large amount of duplication. If we're really talking about a large amount of duplication, then `get_or_create` stops after the first `SELECT`, and there's not much gained from a hash solution.